### PR TITLE
ISPRK-3 Infinispan Query DSL from my Spark application

### PR DESCRIPTION
**Pull request by ghost**
https://issues.jboss.org/browse/ISPRK-3

Added a couple of new operators to the InfinispanRDD:

``` scala
val rdd = new InfinispanRDD[Integer, User](...)

//FilterByQuery
val query = Search.getQueryFactory(cache).from(classOf[User])
                           .having("name").equal("jerry").toBuilder[RemoteQuery].build

val filtered = rdd.filterByQuery(query, classOf[User]).sortByKey().first

// filterByCustom
val filtered = rdd.filterByCustom("my-deployed-filter-factory").filter(_.age > 10).collect()
```

### CODE REVIEW

**Comment by ghost** [comment]
Updated with support for the Java API and more tests covering deployed filters with parameters.

File: src/main/scala/org/infinispan/spark/rdd/FilteredInfinispanRDD.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      if (filterFactory.isDefined) super.count()
+      else
+         filterQuery.fold(parent.remoteCache.size)(_.queryObj.getResultSize)
+   }
+
+   @DeveloperApi
+   override def compute(split: Partition, context: TaskContext): Iterator[(K, R)] = {
+      val (factory, params) = filterQuery.fold((filterFactory.orNull, filterParams))(q => (Filters.ITERATION_QUERY_FILTER_CONVERTER_FACTORY_NAME, Array(q.query)))
+      parent.compute(split, context, address => {
+         val cm = new RemoteCacheManager(createBuilder(address).build())
```
Comment (galderz) [inline src/main/scala/org/infinispan/spark/rdd/FilteredInfinispanRDD.scala:47]: What's the lifecycle of this `RemoteCacheManager` like? IOW, are you making sure it gets closed properly? `RemoteCacheManager` holds resources like connections pools, thread pools, so you don't want to leak those.

File: src/main/scala/org/infinispan/spark/rdd/InfinispanRDD.scala [hunk 2]
```diff
@@ -20,30 +22,37 @@ class InfinispanRDD[K, V](@transient val sc: SparkContext,
...
-      val remoteCacheManager = new RemoteCacheManager(new ConfigurationBuilder().withProperties(configuration).pingOnStartup(true).build())
+   private[rdd] def createBuilder(address: InetSocketAddress) = {
+      new ConfigurationBuilder().withProperties(configuration)
+              .addServer().host(address.getHostString).port(address.getPort)
+              .balancingStrategy(new PreferredServerBalancingStrategy(address))
+   }
+
+   @transient lazy val remoteCache: RemoteCache[K, V] = {
+      val config = new ConfigurationBuilder().withProperties(configuration).pingOnStartup(true).build()
+      val remoteCacheManager = new RemoteCacheManager(config)
```
Comment (galderz) [inline src/main/scala/org/infinispan/spark/rdd/InfinispanRDD.scala:33]: Same story about this `RemoteCacheManager` instance

File: src/main/scala/org/infinispan/spark/rdd/InfinispanRDD.scala [hunk 3]
```diff
@@ -20,30 +22,37 @@ class InfinispanRDD[K, V](@transient val sc: SparkContext,
...
+      val remoteCacheManager = cacheManagerProvider(address)
       val cache = Option(cacheName).map(name => remoteCacheManager.getCache(name)).getOrElse(remoteCacheManager.getCache)
       val segmentFilter = infinispanPartition.segments.map(setAsJavaSet).orNull
-      val filterFactory = config.read[String](InfinispanRDD.FilterFactory)
-      val closeableIterator = cache.retrieveEntries(filterFactory.getOrElse(null.asInstanceOf[String]), segmentFilter, batch)
+      val closeableIterator = cache.retrieveEntries(filterFactory, factoryParams, segmentFilter, batch)
+
       context.addTaskCompletionListener(t => {
          closeableIterator.close()
          remoteCacheManager.stop()
```
Comment (galderz) [inline src/main/scala/org/infinispan/spark/rdd/InfinispanRDD.scala:58]: Is this where you make sure remote cache manager gets stopped, right? Now that I see where you stop the remote cache manager, how often are you going to be creating/stopping remote cache managers? Wondering if some caching could be done? Again, connection pools start/stop, thread pools...etc.

**Comment by galderz** [comment]
The rest looks good

**Comment by ghost** [comment]
@galderz A cache manager is created per node where the InfinispanRDD is executed, and closed in the listeners as you've seen it. 
Wrt caching, it's cached when possible: there's a `lazy val` that stores the RemoteCache instance that preserves the underlying cache manager across calls in the same node. But since Spark actually serializes and sends the RDD to the remote workers, a new CM needs to be created to iterate the segments.

Btw, there was a typo in one of the listeners: it was calling `remoteCache.stop` instead of `remoteCacheManager.stop`, fixed and re-pushed

**Comment by galderz** [comment]
Ok, makes sense. Integrating...

**Comment by galderz** [comment]
Integrated

**Pull request closed without merging**